src/ui/ship_building/ship_building_view.gd

```diff
@@ -27,6 +34,38 @@ func _ready():
 	get_node("UI").visible = true
 
 
+func _physics_process(_delta):
+	# We can only call get_world_2d().direct_space_state safely 
+	# in the physics frame, hence why this logic is here 
+	# instead of _input.
+	if Input.is_action_just_pressed("cancel_place_building"):
+		# Don't exit build view if we have a building preview active
+		if $ShipSprite/ShipGrid.current_building:
+			return
+		# Check if the mouse is over a building collider
+		var space = get_world_2d().direct_space_state
+		var params = PhysicsPointQueryParameters2D.new()
+		params.position = get_global_mouse_position()
+		params.collide_with_areas = true
+		# Only detect physics layer 2 - buildings
+		params.collision_mask = 0b0010
+		# Check if there is a collision at the mouse position
+		if space.intersect_point(params, 1):
+			return
+		
+		emit_signal("ship_deselected")
+
+
+func _input(event):
+	if event is InputEventMouseButton:
+		# Enter build menu when we left click the ship
+		if event.button_index == MOUSE_BUTTON_LEFT and \
+		event.pressed:
+			if is_ship_hover:
+				emit_signal("ship_selected")
+				$ShipSprite.modulate = ship_no_highlight
```

Comment:
Collision layer changes are to facilitate the clicking/right-clicking of the ship sprite to enter/exit build mode.

When the build grid is open we can right click to exit it, but we don't want to do this in a few cases - namely:
when we have a building preview selected
when we're hovering over an existing building
This is because in these cases we want a different behaviour for the right click, so we don't want to exit the build grid.

To facilitate this I've added the following code in the physics step of ship_building_view.gd, we check whether the mouse connects with anything and if it connects with a building physics object - we don't exit the build grid.

Because the empty grid tile also has collision, and we also want to exit if the player right clicks within the empty buildable grid area, we need separate collision layers for empty-but-buildable space, and buildings. The raycast code checks for collisions with the buildings layer only.

As it stands on that branch now these are the physics layers in use:

    1: Empty build grid/generic collisions we don't really care about
    2: Buildings
    3: Warehouse effect range